src/MatrixFields/matrix_multiplication.jl

```diff
@@ -0,0 +1,338 @@
+Replacing the variable idx′ with the variable prod_d = idx′ - idx gives us
+    (M1 ⋅ M2)[idx, idx + prod_d] =
+        ∑_{
+            idx′′ ∈
+                max(left_idx, idx + ld1, idx + prod_d - ud2):
+                min(right_idx, idx + ud1, idx + prod_d - ld2)
+        } M1[idx, idx′′] * M2[idx′′, idx + prod_d].
+Replacing the variable idx′′ with the variable d = idx′′ - idx gives us
+    (M1 ⋅ M2)[idx, idx + prod_d] =
+        ∑_{
+            d ∈
+                max(left_idx - idx, ld1, prod_d - ud2):
+                min(right_idx - idx, ud1, prod_d - ld2)
+        } M1[idx, idx + d] * M2[idx + d, idx + prod_d].
+This can be rewritten using the standard indexing notation as
+    (M1 ⋅ M2)[idx][prod_d] =
+        ∑_{
+            d ∈
+                max(left_idx - idx, ld1, prod_d - ud2):
+                min(right_idx - idx, ud1, prod_d - ld2)
+        } M1[idx][d] * M2[idx + d][prod_d - d].
+Finally, we can express this in terms of left/right boundaries and an interior:
+    (M1 ⋅ M2)[idx][prod_d] =
+        ∑_{
+            d ∈
+                if idx < left_idx - ld1
+                    max(left_idx - idx, prod_d - ud2):min(ud1, prod_d - ld2)
+                elseif idx > right_idx - ud1
+                    max(ld1, prod_d - ud2):min(right_idx - idx, prod_d - ld2)
+                else
+                    max(ld1, prod_d - ud2):min(ud1, prod_d - ld2)
+                end
+        } M1[idx][d] * M2[idx + d][prod_d - d].
+
+We only need to define (M1 ⋅ M2)[idx][prod_d] when it has a nonzero value in the
+interior, which will be the case when
+    max(ld1, prod_d - ud2) ≤ min(ud1, prod_d - ld2).
+This can be rewritten as a system of four inequalities:
+    ld1 ≤ ud1,
+    ld1 ≤ prod_d - ld2,
+    prod_d - ud2 ≤ ud1, and
+    prod_d - ud2 ≤ prod_d - ld2.
+By definition, ld1 ≤ ud1 and ld2 ≤ ud2, so the first and last inequality are
+always true. Rearranging the remaining two inequalities gives us
+    ld1 + ld2 ≤ prod_d ≤ ud1 + ud2.
+=#
+
+struct TopLeftMatrixCorner <: BoundaryCondition end
+struct BottomRightMatrixCorner <: BoundaryCondition end
+
+has_boundary(
+    ::MultiplyColumnwiseBandMatrixField,
+    ::LeftBoundaryWindow{name},
+) where {name} = true
+has_boundary(
+    ::MultiplyColumnwiseBandMatrixField,
+    ::RightBoundaryWindow{name},
+) where {name} = true
+
+get_boundary(
+    ::MultiplyColumnwiseBandMatrixField,
+    ::LeftBoundaryWindow{name},
+) where {name} = TopLeftMatrixCorner()
+get_boundary(
+    ::MultiplyColumnwiseBandMatrixField,
+    ::RightBoundaryWindow{name},
+) where {name} = BottomRightMatrixCorner()
+
+stencil_interior_width(::MultiplyColumnwiseBandMatrixField, matrix1, arg) =
+    ((0, 0), outer_diagonals(eltype(matrix1)))
+
+function boundary_width(
+    ::MultiplyColumnwiseBandMatrixField,
+    ::TopLeftMatrixCorner,
+    matrix1,
+    arg,
+)
+    ld1, _ = outer_diagonals(eltype(matrix1))
+    return max((left_idx(axes(arg)) - ld1) - left_idx(axes(matrix1)), 0)
+end
```

Comment:
I've changed how this works in #1322 which simplifies the logic, can you use that approach?